include/glfw/glfw/input/mouse.h

```diff
@@ -0,0 +1,110 @@
+#pragma once
+
+#include "glfw/core.h"
+
+namespace glfw::input
+{
+
+using MouseButtonIndex = int;
+
+struct MouseMovement
+{
+    double dx = 0.0;
+    double dy = 0.0;
+};
+
+struct MousePosition
+{
+    double x = 0.0;
+    double y = 0.0;
+
+  public:
+    friend MouseMovement operator-( const MousePosition& lhs, const MousePosition& rhs )
+    {
+        return MouseMovement{ lhs.x - rhs.x, lhs.y - rhs.y };
+    }
+};
+
+class MouseHandler
+{
+  private:
+    static void buttonCallbackWrapper( GLFWwindow* window, int key, int action, int mods )
+    {
+        instance( window ).buttonCallback( key, static_cast<ButtonAction>( action ), ModifierFlag{ mods } );
+    }
+
+    void buttonCallback( MouseButtonIndex key, ButtonAction action, ModifierFlag modifier )
+    {
+        assert( action != ButtonAction::e_repeat );
+
+        {
+            auto g = std::lock_guard{ m_mx };
+            m_event_map[ key ].pushEvent( ButtonEvent{ .mods = modifier, .action = action } );
+        }
+    }
+
+  private:
+    MouseHandler( GLFWwindow* window )
+        : m_window{ window }
+    {
+        glfwSetMouseButtonCallback( window, buttonCallbackWrapper );
+        m_position = getPosition();
+    }
+
+    MousePosition updatePosition() { return std::exchange( m_position, getPosition() ); }
+
+  public:
+    using ButtonEventMap = std::unordered_map<MouseButtonIndex, ButtonEventInfo>;
+    static MouseHandler& instance( GLFWwindow* window )
+    {
+        return *s_handler_table.lookup( window, [ window ]() {
+            return std::unique_ptr<MouseHandler>{ new MouseHandler{ window } };
+        } );
+    }
+
+  public:
+    void setHidden() { glfwSetInputMode( m_window, GLFW_CURSOR, GLFW_CURSOR_HIDDEN ); }
+    void setNormal() { glfwSetInputMode( m_window, GLFW_CURSOR, GLFW_CURSOR_NORMAL ); }
+
+  private:
+    MousePosition getPosition()
```

Comment:
It is const method.

Reply:
Hm, I wasn't sure whether it's semantically const